bug: ensure_equals now works correctly with floating numbers

remileduc opened the pull request:
To compare 2 floating points, we need to take care of the epsilon of the machine.
On top of that, the precision of a double depends on the length of the radix (the bigger it is, the less information can be stored in the floating part).

Thus, I've updated the comparison between 2 floating points values according to http://en.cppreference.com/w/cpp/types/numeric_limits/epsilon (see the example code).

- fix comparison between doubles when the radix is big
- fix comparison between long doubles (before they were cast in double with possible wrong comparison)
    - this is now a templated function

I've updated the tests, I'm not quiet familiar yet with this library so don't hesitate to correct my code or tell me if I'm doing wrong stuff...

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -130,12 +130,16 @@ void ensure_equals(const M& msg, const LHS * const actual, const RHS * const exp
-void ensure_equals(const M& msg, const double& actual, const double& expected, const double& epsilon)
+template<typename M, typename T>
+typename std::enable_if<std::is_floating_point<T>::value, void>::type
+    ensure_equals(const M& msg, const T& actual, const T& expected, const T& epsilon)
```
This is the C++11 version.

The C++14 version would be `std::is_floating_point<T>()`

The C++17 version would be `std::is_floating_point_v<T>`

mrzechonek replied:
As far as I know, `std::is_floating_point<T>::value` is still valid and the two forms you mention are just syntactic sugar?

mrzechonek replied:
I think you're missing `#include <type_traits>` before using `std::enable_if`. Also, this needs to be properly wrapped in `#if __cplusplus >= ...` clause or rewritten to work in C++98 mode.

remileduc replied:
True, I use MSVC at work, which is quite permissive on STL includes...

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -130,12 +130,16 @@ void ensure_equals(const M& msg, const LHS * const actual, const RHS * const exp
-
-    if ( (actual != expected) && !((diff <= epsilon) && (diff >= -epsilon )) )
+	if (std::isnan(actual) && std::isnan(expected))
+		return;
```
Usually, `NaN != NaN`, but in a test library it may be better to have `NaN == NaN`, what do you think?

I have the case where I want to assert that a value is `NaN`, so I have 2 options:

```c++
ensure_not(myNaN == myNaN); // current version
ensure_equals(myNaN, quiet_NaN(); // if we use the suggested code
```

mrzechonek replied:
I'd rather have `ensure_isnan(value)` instead of abusing comparison rules.

It might even be a good idea to add generic `ensure_call(<function>, <arguments...>)`, provided we're using a recent standard.

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -149,10 +153,11 @@ void ensure_equals(const M& msg, const double& actual, const double& expected, c
+    ensure_equals(const M& msg, const T& actual, const T& expected)
 {
-    ensure_equals(msg, actual, expected, std::numeric_limits<double>::epsilon());
+    ensure_equals(msg, actual, expected, std::numeric_limits<T>::epsilon() * std::abs(actual + expected));
```
here we scale the epsilon regarding the length of the radix

remileduc commented on selftest/ensure_equals.cpp:
```diff
@@ -147,9 +147,9 @@ template<>
-    lhs /= 2;
+    double lhs = 1000000;
+    double rhs = 200000;
+    lhs = lhs / std::sqrt(5) / std::sqrt(5);
```
This example doesn't pass in master.

mrzechonek commented on selftest/ensure_equals.cpp:
```diff
@@ -177,11 +177,33 @@ void object::test<14>()
 template<>
 template<>
-void object::test<15>()
+void object::test<16>()
```
Please add new tests at the end.

mrzechonek commented:
Overall this looks like a good idea, but the implementation must be backward compatible - I know there are people using TUT with quite old/peculiar compilers, especially for embedded.

jbakosi commented:
Yes, please keep TUT max C++11 for now if you guys can. Thanks.

remileduc commented:
I've updated the PR, tests should pass...

I have updated the code so we can mix float and double in `ensure_equals`, however it adds some code complexity. I'm not an expert with templates, so maybe I don't use the right method (I put some comments to explain what I intend to do).

I also have a question: is it a good idea to use the `__cplusplus` value? MSVC has it hardcoded to `199711` (even in the 2017 version), and from the tests, it seems that only G++6 takes advantage of it (all the other tests are skipping the the new test 21).
Which means that to use TUT correctly, most of us will have to remove these checks manually... (I don't know if there is a better solution though)

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -92,7 +96,12 @@ void ensure_not(const M& msg, bool cond)
+#else
+void
+#endif
+    ensure_equals(const M& msg, const LHS& actual, const RHS& expected)
```
Maybe not the most beautiful way to do it... same below.
Feel free to comment :)

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -130,12 +139,22 @@ void ensure_equals(const M& msg, const LHS * const actual, const RHS * const exp
+    ensure_equals(const M& msg, const LHS& actual, const RHS& expected, const EPS& epsilon)
+{
+	if((actual != expected)
+		&& (std::isnan(actual) || std::isnan(expected)
```
We need to keep this tests, otherwise `std::abs()` will complain if it gets a NaN value.
We now have a consistent behavior with the standard: if one or both values are NaN, the test fails.

remileduc commented on include/tut/tut_assert.hpp:
```diff
@@ -149,10 +168,20 @@ void ensure_equals(const M& msg, const double& actual, const double& expected, c
+typename std::enable_if<std::is_floating_point<LHS>::value || std::is_floating_point<RHS>::value, void>::type
+    ensure_equals(const M& msg, const LHS& actual, const RHS& expected)
+{
+	auto epsilon = sizeof(LHS) < sizeof(RHS) ? std::numeric_limits<LHS>::epsilon() : std::numeric_limits<RHS>::epsilon();
```
The goal is here to get the biggest epsilon if `LHS` and `RHS` are not of the same type

remileduc commented on selftest/ensure_equals.cpp:
```diff
@@ -379,5 +379,24 @@ void object::test<20>()
+{
+#if __cplusplus < 201103L
+    skip();
+#endif
```
If we don't use the new template function, compiler will resolve `double` + `float` to the most generic function because there is no conversion to do. Thus, it won't use the `double` + `double` version of `ensure_equals` (because of the conversion `float` to `double`), and it will try to compare the values like an `int` with a simple `!=`.

Thus the `skip()` if we can't use the new template function.

remileduc commented:
Are you still interested? How can I improve it so it gets merged?

remileduc commented:
I just answer myself about my comment about MSVC and the `__cplusplus` variable : it is now possible to have it right!!! https://blogs.msdn.microsoft.com/vcblog/2018/04/09/msvc-now-correctly-reports-__cplusplus/

I wonder: would you be more interested if, instead of having test of this variable everywhere, I create a /c++11/ header file and we include the correct header file depending on the `__cplusplus`?

mrzechonek commented:
Hi. Sorry about radio silence.

Yes, I am still interested in merging this PR, but recently I have very little time to work on side projects :(

1. about a separate header: I'd rather have #ifdefs, as in the current version.
2. about improvements: after adding `ensure_isnan` (+self test) and fixing test ordering, I think it's ready for merging

remileduc commented:
Hi,

No problem, I had quite some work lately.

In my last commit, I've added the following methods, along with a new test file to test them:

- `ensure_isnan`
- `ensure_not_isnan`
- `ensure_isinf`
- `ensure_isfinite`

Note that they follow the C++ standard, no other tests are performed (like check that the type `has_inf`...)

Is it OK for you?

mrzechonek commented:
Thank you!
